### DISCUSSION TITLE
How to use variables within exec pipelines

### DISCUSSION THREAD

**Discussion by tobrien-nydig** [helmfile/helmfile#184] [Questions]
I'm trying to automatically generate releases based on folder structures and I'm running into a hiccup here.

It looks like my `$team` variable is coming up as empty when it's being run. When I hard code in the team name it works just fine. Any ideas on what I'm doing wrong?

```
{{- $team := env "ARGOCD_ENV_team" }}
releases:
# This loops through the current directory and makes a release for each directory's name
{{- range $_, $file := ( exec "bash" (list "-c" "echo -n ../values/`{{$team}}`/*/ | sed 's#../values/`{{$team}}`/##g' | sed 's#/##g'") | splitList " " ) }}
  - name: "{{ $file }}"
    <<: *default
{{ end }}
```

**Comment by philomory** [comment]
I don't think the syntax you are using is correct; to the best of my knowledge, inside a go text template segment, something like ``"foo`{{$team}}`bar"`` is just a string literal containing the word `foo`, a backtick character, two left curly braces, a dollar sign, the word `team`, to right curly braces, another backtick, and the word `bar`. There's no special characters being interpreted.

Thus, you are effectively running the following shell command: 
```
echo -n ../values/`{{$team}}`/*/ | sed 's#..values/`{{$team}}`/##g' | sed 's#/##g'
```
The first thing that happens at that point is that bash tries to perform its own variable substitution, although `$team` is not defined, resulting in (for just the first pipeline segment): ``echo -n ../values/`{{}}`/*/``. Bash then sees the `` `{{}}` `` portion, and takes the backticks as a cue to perform command substitution, much as if it were instead `echo -n ../values/$({{}})/*`. However, `{{}}` is not a valid command, so bash just substitute the empty string (normally bash would also print a warning to stderr here, but `helmfile` swallows such output, even in debug mode. You can see it by running something like ``bash -c 'echo -n ../values/`{{$team}}`/*/'`` from your shell, though).

I think you may have been conflating `` `{{ ... }}` `` with ``{{` ... `}}``, which _is_ meaningful syntax in Go templates, though it still doesn't really do what you want.

One viable option would be to simply use the `envExec` function rather than plain `exec`, something like this:
```
{{- range $_, $file := ( envExec (dict "team" $team) "bash" (list "-c " "echo -n ../values/$team/*/ | sed \"s#../values/$team/##g\" | sed 's#/##g'") | splitList " ") }}
```
... although I admit, I'm still not sure exactly what this is meant to accomplish, since if `$team` is, say, `foo`, this will literally be `echo -n ../values/foo/*/ | sed "s#../values/foo/##g" | sed 's#/##g'`, which is just going to going to literally be the string `*`, every single time.

Either way, it may be easier to write shell script that takes the team as an argument and outputs the text you want, and the invoke it via something like
```
{{- range $_, $file := exec "./scripts/myscript.sh" (list $team) }}
```
... since this is both simpler, and easier to test.

**Reply by philomory** [reply]
If you're trying to find a list of files in the folder `../values/$team`, such a script might look like this:

```bash
#!/usr/bin/bash

for f in ../values/$1/* ;
do
  basename $f
done
```
... although you'd then want to use `| trim | splitList "\n"` instead of `| splitList " "`

**Reply by philomory** [reply]
Perhaps another, simpler command to accomplish what I think you are after would be simply `ls -1 ../values/$team/* | xargs -n1 basename`.
